### DISCUSSION TITLE
Remaining loaders/savers to rework for the source/target API

### DISCUSSION THREAD

**Discussion by kleisauke** [libvips/libvips#2051] [Ideas]
This issue tracks the remaining loaders and savers that might be reworked for the new [source/target API](https://libvips.github.io/libvips/2019/11/29/True-streaming-for-libvips.html). I mostly used [this comment](https://github.com/libvips/libvips/pull/1443#issuecomment-552190631) as reference. Feel free to add more information in the comments.

| Image suffix(es)       | Loader                                    | Saver                   | Finished                           | 
|------------------------|-------------------------------------------|-------------------------|:----------------------------------:|
| -                      | [rawload]                                 | [rawsave]<sup>1</sup>   | Loader 🚧<br/>Saver ✅             |
| .v, .vips              | [vipsload]<sup>2</sup>                    | [vipssave]<sup>2</sup>  | Loader ⚠️<br/>Saver ⚠️             |
| [whole-slide suffixes] | [openslideload]                           | -                       | Loader ⚠️                          |
| .fits                  | [fitsload]<sup>2</sup>                    | [fitssave]<sup>2</sup>  | Loader ⚠️<br/>Saver 🚧             |
| [nifti suffixes]       | [niftiload]<sup>2</sup>                   | [niftisave]<sup>2</sup> | Loader ⚠️<br/>Saver 🚧             |
| [IM suffixes]          | [magickload (IM6)]<br/>[magickload (IM7)] | [magicksave]            | Loader 💡<sup>3</sup><br/>Saver 🚧 |

Loaders and savers missing from the above table: [analyzeload], [matload], [openexrload].

### Notes:
1. See PR https://github.com/libvips/libvips/pull/3942.
2. We'd get early close for free if this is reworked.
3. The source is only loaded into memory if it's not file-based, see PR https://github.com/libvips/libvips/pull/4592.

<details>
  <summary>Already supports the new source/target API</summary>

| Image suffix(es)       | Loader                                     | Saver                                      | Finished                           | 
|------------------------|--------------------------------------------|--------------------------------------------|:----------------------------------:|
| .jpg, .jpeg, .jpe      | [jpegload]                                 | [jpegsave]                                 | Loader ✅<br/>Saver ✅             |
| .png                   | [pngload (libpng)]<br/>[pngload (libspng)] | [pngsave (libpng)]<br/>[pngsave (libspng)] | Loader ✅<br/>Saver ✅             |
| .webp                  | [webpload]                                 | [webpsave]                                 | Loader 💡<br/>Saver ✅             |
| .gif                   | [gifload]<sup>5</sup>                      | [gifsave]<sup>6</sup>                      | Loader 💡<br/>Saver ✅             |
| .heif, .heic, .avif    | [heifload]<sup>7</sup>                     | [heifsave]<sup>7</sup>                     | Loader ✅<br/>Saver ✅             |
| .jxl                   | [jxlload]                                  | [jxlsave]                                  | Loader ✅<br/>Saver ✅             |
| [JP2K suffixes]        | [jp2kload]                                 | [jp2ksave]                                 | Loader ✅<br/>Saver ✅             |
| .pdf                   | [pdfload (Poppler)]<br/>[pdfload (PDFium)] | -                                          | Loader ✅                          |
| .svg, .svgz, .svg.gz   | [svgload]                                  | -                                          | Loader ✅                          |
| .csv                   | [csvload]                                  | [csvsave]                                  | Loader ✅<br/>Saver ✅             |
| .ppm, .pgm, .pbm, .pfm | [ppmload]                                  | [ppmsave]                                  | Loader 💡<sup>4</sup><br/>Saver ✅ |
| .mat                   | [matrixload]<sup>8</sup>                   | [matrixsave]<sup>8</sup>                   | Loader ✅<br/>Saver ✅             |
| .hdr                   | [radload]                                  | [radsave]                                  | Loader ✅<br/>Saver ✅             |
| .tif, .tiff            | [tiffload]                                 | [tiffsave]<sup>9</sup>                     | Loader ✅<br/>Saver ✅             |
| .dz                    | -                                          | [dzsave]<sup>10</sup>                      | Saver ✅                           |

### Notes:
4. The source is only loaded into memory if it can be efficiently memory-mapped and the PPM is binary (with a bit depth of at least 8 bits).
5. libnsgif needs the whole source file available at once, it can't stream images (as far as I know), so it's not a great fit. It's rather like webp. giflib isn't much better: you can only find out the number of frames by scanning the entire image, and you have to discover the number of frames as part of reading the header.
6. See upstream PR https://github.com/dloebl/cgif/pull/7.
7. We could use the [heif_context_read_from_reader] and [heif_context_write] functions.
8. [matrixload] / [matrixsave] are used to read and write convolution matrices. They are very like the CSV load/save pair.
9. See PR https://github.com/libvips/libvips/pull/2798.
10. See PR https://github.com/libvips/libvips/pull/2718.
</details>

Legend:
| Symbol  | Meaning                                                      |
|:-------:|:------------------------------------------------------------:|
| ✅      | Source/target support available                              |
| 💡      | Needs to load the entire source into memory                  |
| ⚠️      | Only if there's a filename associated with the source/target |
| 🚧      | Source/target support not (yet) available                    |

[analyzeload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/analyzeload.c
[csvload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/csvload.c
[csvsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/csvsave.c
[dzsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/dzsave.c
[fitsload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/fitsload.c
[fitssave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/fitssave.c
[gifload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/nsgifload.c
[gifsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/cgifsave.c
[heifload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/heifload.c
[heifsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/heifsave.c
[jp2kload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jp2kload.c
[jp2ksave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jp2ksave.c
[jpegload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jpegload.c
[jpegsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jpegsave.c
[jxlload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jxlload.c
[jxlsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/jxlsave.c
[magickload (IM6)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/magick6load.c
[magickload (IM7)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/magick7load.c
[magicksave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/magicksave.c
[matload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/matload.c
[matrixload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/matrixload.c
[matrixsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/matrixsave.c
[niftiload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/niftiload.c
[niftisave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/niftisave.c
[openexrload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/openexrload.c
[openslideload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/openslideload.c
[pdfload (PDFium)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/pdfiumload.c
[pdfload (Poppler)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/popplerload.c
[pngload (libpng)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/pngload.c
[pngload (libspng)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/spngload.c
[pngsave (libpng)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/pngsave.c
[pngsave (libspng)]: https://github.com/libvips/libvips/blob/master/libvips/foreign/spngsave.c
[ppmload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/ppmload.c
[ppmsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/ppmsave.c
[radload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/radload.c
[radsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/radsave.c
[rawload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/rawload.c
[rawsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/rawsave.c
[svgload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/svgload.c
[tiffload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/tiffload.c
[tiffsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/tiffsave.c
[vipsload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/vipsload.c
[vipssave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/vipssave.c
[webpload]: https://github.com/libvips/libvips/blob/master/libvips/foreign/webpload.c
[webpsave]: https://github.com/libvips/libvips/blob/master/libvips/foreign/webpsave.c

[whole-slide suffixes]: https://github.com/libvips/libvips/blob/babbb086f92a678b0aad87826d8ab738bce5570f/libvips/foreign/openslideload.c#L306-L315
[nifti suffixes]: https://github.com/libvips/libvips/blob/b0b8e6ee65db6ff2a6a45f9f7b79cf7dbc8f4398/libvips/foreign/niftiload.c#L634-L640
[IM suffixes]: https://imagemagick.org/script/formats.php#supported
[JP2K suffixes]: https://github.com/libvips/libvips/blob/b0b8e6ee65db6ff2a6a45f9f7b79cf7dbc8f4398/libvips/foreign/jp2kload.c#L943-L944

[heif_context_read_from_reader]: https://github.com/strukturag/libheif/blob/f3e58e21a87be3704c5c4837984be7eb235717d5/libheif/heif.h#L395-L399
[heif_context_write]: https://github.com/strukturag/libheif/blob/f3e58e21a87be3704c5c4837984be7eb235717d5/libheif/heif.h#L902-L905

**Comment by jcupitt** [comment]
Oh hey well done! This is a great idea.

I've started revising gifload to work with streams:

https://github.com/libvips/libvips/tree/revise-gifload2

not done yet though.

**Comment by jcupitt** [comment]
OK, gifload done and merged to master.

**Comment by kleisauke** [comment]
Awesome! I confirm `gifload_source` is working well. Tested with:
```bash
wget https://storage.googleapis.com/downloads.webmproject.org/webp/images/dancing-banana.gif
cat dancing-banana.gif | vips copy stdin[n=-1] x.webp
```

I'll have a look at `pdfload` this week (will be quite similar to `svgload`, I think).

**Comment by jcupitt** [comment]
I did csv load and save as well, and added `csvload_source` and `csvsave_target`.

**Comment by jcupitt** [comment]
`matrixload` and `matrixsave` done too, plus `_source` versions.

**Comment by jcupitt** [comment]
`pdfload` done, and I added `pdfload_source`.

**Comment by jcupitt** [comment]
`heifload` and `heifsave` done too, plus `_source`/`_target` versions.

**Comment by kleisauke** [comment]
I started to revise `vipsload` for the source API, see: https://github.com/kleisauke/libvips/commit/a11b978fb70d338c2d7ffba8fd2b4ff1106d1b87.

While doing that, I noticed something strange in `csvload_source` and `matrixload_source`. I suspect that these lines: https://github.com/libvips/libvips/blob/175408733fcffed7aa0618c6d5ad5390ed5c4f88/libvips/foreign/matrixload.c#L405
https://github.com/libvips/libvips/blob/175408733fcffed7aa0618c6d5ad5390ed5c4f88/libvips/foreign/csvload.c#L603

Should check for `source->source` instead.

**Comment by jcupitt** [comment]
Wow you're right, they are all messed up. Too much copy-paste programming :(

I've revised them. Thanks!

**Comment by jcupitt** [comment]
... I'll try to add some tests too.

**Comment by jcupitt** [comment]
OK, tests added, plus ppm load/save to/from a connection.

**Comment by kleisauke** [comment]
I've added `openslideload` to the list and marked it as finished, since this is most likely done with commit https://github.com/libvips/libvips/commit/babbb086f92a678b0aad87826d8ab738bce5570f.

If someone wants to work on revising `vipsload`, then see https://github.com/libvips/libvips/issues/1541#issuecomment-649766008. Otherwise, I'll pick this up in the course of next week.

**Comment by jcupitt** [comment]
I'll add some tests for openslideload_source().

I'm reworking vipsdisp to use the new Source API, hence my current interest in this:

https://github.com/jcupitt/vipsdisp

**Comment by jcupitt** [comment]
I did vipsload_source() and updated the table.

**Comment by jcupitt** [comment]
... and vipssave_target.

**Comment by jcupitt** [comment]
I did niftiload and fitsload. The savers still need doing.

**Comment by jcupitt** [comment]
... tiffsave and dzsave now done.

**Reply by kleisauke** [reply]
Great! I updated the start post to mention those PRs and will do some testing tomorrow.

FWIW, the weserv/images project could use the source/target API once `magickload_source` is added (to ensure that ICO, BMP and PSD images can still be loaded). See:
https://github.com/weserv/images/blob/05563d31959905ed9039b9bb769658e17241d344/src/api/CMakeLists.txt#L79-L85
https://github.com/weserv/images/blob/05563d31959905ed9039b9bb769658e17241d344/test/api/base.cpp#L15-L16

**Reply by jcupitt** [reply]
The tiffsaver is a bit disappointing -- I hadn't realized that libtiff has to be able to read from the output. It'll write a chunk of data, write another chunk, then rewind and adjust some pointers. This means you have to buffer the whole output object in memory before you can start to write to a pipe :( 

It's fine for filesystem output, of course.

**Reply by kleisauke** [reply]
I think buffering the entire output object into memory is OK for non-file based targets. Perhaps we could use `TIFFFdOpen()` for file-based targets and `TIFFClientOpen()` for memory-based targets ([similar to Pillow](https://github.com/python-pillow/Pillow/blob/ef8cec6185db64ad501a191ffcae9476b55aea72/src/libImaging/TiffDecode.c#L770-L798))? This would avoid having the `->read` and `->seek` in the target class, which is somewhat confusing.

**Reply by jcupitt** [reply]
Yes, it's weird isn't it. I don't think anyone would design a format now that could only be written by something that could read its own output.

I wanted to make it work for subclasses too, so more than just memory and file. You only need to implement seek and read if you need tiff save support, so it's not too much of a burden, I hope.

**Reply by kleisauke** [reply]
> I wanted to make it work for subclasses too, so more than just memory and file. You only need to implement seek and read if you need tiff save support, so it's not too much of a burden, I hope.

Ah OK, seems reasonable to me. I had a quick attempt to implement this in NetVips and the weserv/images project, see commits:
https://github.com/kleisauke/net-vips/commit/54a04c837c4db5e9cc8a37263e9c68cd030374bc
https://github.com/weserv/images/commit/fabfb0911ae7ff94ef5bda65e88209c3e1906314

It works great! :tada:

**Reply by jcupitt** [reply]
Great! It's in master now. I should test it in pyvips and ruby-vips too.
